clang/lib/CodeGen/CGRecordLayoutBuilder.cpp

```diff
@@ -349,10 +349,11 @@ void CGRecordLowering::lowerUnion() {
     if (!IsZeroInitializable)
       continue;
     // Conditionally update our storage type if we've got a new "better" one.
-    if (!StorageType ||
-        getAlignment(FieldType) >  getAlignment(StorageType) ||
+    if (!StorageType || getAlignment(FieldType) > getAlignment(StorageType) ||
         (getAlignment(FieldType) == getAlignment(StorageType) &&
-        getSize(FieldType) > getSize(StorageType)))
+         getSize(FieldType) > getSize(StorageType)) ||
+        (DataLayout.isFatPointer(FieldType) &&
+         !DataLayout.isFatPointer(StorageType)))
```

Comment:
This works for the simple case but not for the more complex case where you have nested structures. Firstly, it needs to be "Field/StorageType contains a fat pointer", but then you run into issues with:
```
union {
    struct {
        int a;
        __intcap b
    } c;
    struct {
        __intcap d;
        int e;
    } f;
};
```
where you need to merge the two types into something isomorphic to a pair of capabilities, though you can always just fall back on an aligned i8 array for complex cases.

Reply:
That is a good point and I'll add a test case for this. Since we don't recurse here my assumption would be that we retain the old logic of using the first struct: `type { i32, i8 addrspace(200)* }`.

Ideally we'd lower it to a two cap struct but this will require further changes.